src/python/pants/engine/build_files.py

```diff
@@ -201,108 +202,99 @@ def _hydrate(item_type, spec_path, **kwargs):
   return item
 
 
+class _MappedSpecs(datatype([
+    ('address_families', tuple),
+    ('specs', Specs),
+])):
+  """Wraps up logic to resolve specs into individual target addresses, or error out."""
+
+  def __new__(cls, address_families, specs):
+    return super(_MappedSpecs, cls).__new__(cls, tuple(address_families), specs)
+
+  @memoized_property
+  def _exclude_compiled_regexps(self):
+    return [re.compile(pattern) for pattern in set(self.specs.exclude_patterns or [])]
+
+  def _excluded_by_pattern(self, address):
+    return any(p.search(address.spec) is not None for p in self._exclude_compiled_regexps)
+
+  @memoized_property
+  def _target_tag_matches(self):
+    def filter_for_tag(tag):
+      return lambda t: tag in [str(t_tag) for t_tag in t.kwargs().get("tags", [])]
+    return wrap_filters(create_filters(self.specs.tags, filter_for_tag))
+
+  def _matches_target_address_pair(self, address, target):
+    return self._target_tag_matches(target) and not self._excluded_by_pattern(address)
+
+  @memoized_property
+  def _address_family_by_directory(self):
+    return {af.namespace: af for af in self.address_families}
+
+  def _matching_addresses_for_spec(self, spec):
+    # NB: if a spec is provided which expands to some number of targets, but those targets match
+    # --exclude-target-regexp, we do NOT fail! This is why we wait to apply the tag and exclude
+    # patterns until we gather all the targets the spec would have matched without them.
+    try:
+      addr_families_for_spec = spec.matching_address_families(self._address_family_by_directory)
+    except Spec.AddressFamilyResolutionError as e:
+      raise ResolveError(e)
```

Comment:
iirc, this will lose a lot of the stack trace from the original exception. We get more descriptive debugging if we use raise from syntax. 

Using the Future library,
```python
from future.utils import raise_from
raise_from(ResolveError(e), e)
```

Although I see after writing this that you’re already passing the original exception. So maybe this isn’t necessary?

Reply:
I think `raise_from` looks like it makes sense here and will use it.

Reply:
Done in a146c99!